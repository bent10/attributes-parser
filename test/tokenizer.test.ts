/// <reference types="vitest/globals" />

import { tokenizeAttrs } from '../src/index.js'

it('should tokenize a string of attributes', () => {
  const input = `id="my-id" class='my-class' num=3.14 numNeg=-3.14 data-num="3.14" data-value="123" data-value=1_000_000 options=\'{"key": "value", "array": [1, 2, 3]}\' data-list="[1, 2, 3]" punc="a=b,c,d,e" checked=false checked=false data-checked="false" disabled`

  const tokens = tokenizeAttrs(input)
  expect(tokens).toMatchSnapshot()
})
