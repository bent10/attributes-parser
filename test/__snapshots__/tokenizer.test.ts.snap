// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should tokenize a string of attributes 1`] = `
[
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 0,
    "text": "id",
    "toString": [Function],
    "type": "AttributeName",
    "value": "id",
  },
  {
    "col": 3,
    "line": 1,
    "lineBreaks": 0,
    "offset": 2,
    "text": "=",
    "toString": [Function],
    "type": "Separator",
    "value": "=",
  },
  {
    "col": 4,
    "line": 1,
    "lineBreaks": 0,
    "offset": 3,
    "text": "\\"my-id\\"",
    "toString": [Function],
    "type": "StringLiteral",
    "value": "my-id",
  },
  {
    "col": 11,
    "line": 1,
    "lineBreaks": 0,
    "offset": 10,
    "text": " ",
    "toString": [Function],
    "type": "WhiteSpace",
    "value": " ",
  },
  {
    "col": 12,
    "line": 1,
    "lineBreaks": 0,
    "offset": 11,
    "text": "class",
    "toString": [Function],
    "type": "AttributeName",
    "value": "class",
  },
  {
    "col": 17,
    "line": 1,
    "lineBreaks": 0,
    "offset": 16,
    "text": "=",
    "toString": [Function],
    "type": "Separator",
    "value": "=",
  },
  {
    "col": 18,
    "line": 1,
    "lineBreaks": 0,
    "offset": 17,
    "text": "'my-class'",
    "toString": [Function],
    "type": "StringLiteral",
    "value": "my-class",
  },
  {
    "col": 28,
    "line": 1,
    "lineBreaks": 0,
    "offset": 27,
    "text": " ",
    "toString": [Function],
    "type": "WhiteSpace",
    "value": " ",
  },
  {
    "col": 29,
    "line": 1,
    "lineBreaks": 0,
    "offset": 28,
    "text": "num",
    "toString": [Function],
    "type": "AttributeName",
    "value": "num",
  },
  {
    "col": 32,
    "line": 1,
    "lineBreaks": 0,
    "offset": 31,
    "text": "=",
    "toString": [Function],
    "type": "Separator",
    "value": "=",
  },
  {
    "col": 33,
    "line": 1,
    "lineBreaks": 0,
    "offset": 32,
    "text": "3.14",
    "toString": [Function],
    "type": "NumericLiteral",
    "value": 3.14,
  },
  {
    "col": 37,
    "line": 1,
    "lineBreaks": 0,
    "offset": 36,
    "text": " ",
    "toString": [Function],
    "type": "WhiteSpace",
    "value": " ",
  },
  {
    "col": 38,
    "line": 1,
    "lineBreaks": 0,
    "offset": 37,
    "text": "numNeg",
    "toString": [Function],
    "type": "AttributeName",
    "value": "numNeg",
  },
  {
    "col": 44,
    "line": 1,
    "lineBreaks": 0,
    "offset": 43,
    "text": "=",
    "toString": [Function],
    "type": "Separator",
    "value": "=",
  },
  {
    "col": 45,
    "line": 1,
    "lineBreaks": 0,
    "offset": 44,
    "text": "-3.14",
    "toString": [Function],
    "type": "NumericLiteral",
    "value": -3.14,
  },
  {
    "col": 50,
    "line": 1,
    "lineBreaks": 0,
    "offset": 49,
    "text": " ",
    "toString": [Function],
    "type": "WhiteSpace",
    "value": " ",
  },
  {
    "col": 51,
    "line": 1,
    "lineBreaks": 0,
    "offset": 50,
    "text": "data-num",
    "toString": [Function],
    "type": "AttributeName",
    "value": "data-num",
  },
  {
    "col": 59,
    "line": 1,
    "lineBreaks": 0,
    "offset": 58,
    "text": "=",
    "toString": [Function],
    "type": "Separator",
    "value": "=",
  },
  {
    "col": 60,
    "line": 1,
    "lineBreaks": 0,
    "offset": 59,
    "text": "\\"3.14\\"",
    "toString": [Function],
    "type": "StringLiteral",
    "value": "3.14",
  },
  {
    "col": 66,
    "line": 1,
    "lineBreaks": 0,
    "offset": 65,
    "text": " ",
    "toString": [Function],
    "type": "WhiteSpace",
    "value": " ",
  },
  {
    "col": 67,
    "line": 1,
    "lineBreaks": 0,
    "offset": 66,
    "text": "data-value",
    "toString": [Function],
    "type": "AttributeName",
    "value": "data-value",
  },
  {
    "col": 77,
    "line": 1,
    "lineBreaks": 0,
    "offset": 76,
    "text": "=",
    "toString": [Function],
    "type": "Separator",
    "value": "=",
  },
  {
    "col": 78,
    "line": 1,
    "lineBreaks": 0,
    "offset": 77,
    "text": "\\"123\\"",
    "toString": [Function],
    "type": "StringLiteral",
    "value": "123",
  },
  {
    "col": 83,
    "line": 1,
    "lineBreaks": 0,
    "offset": 82,
    "text": " ",
    "toString": [Function],
    "type": "WhiteSpace",
    "value": " ",
  },
  {
    "col": 84,
    "line": 1,
    "lineBreaks": 0,
    "offset": 83,
    "text": "data-value",
    "toString": [Function],
    "type": "AttributeName",
    "value": "data-value",
  },
  {
    "col": 94,
    "line": 1,
    "lineBreaks": 0,
    "offset": 93,
    "text": "=",
    "toString": [Function],
    "type": "Separator",
    "value": "=",
  },
  {
    "col": 95,
    "line": 1,
    "lineBreaks": 0,
    "offset": 94,
    "text": "1_000_000",
    "toString": [Function],
    "type": "NumericLiteral",
    "value": 1000000,
  },
  {
    "col": 104,
    "line": 1,
    "lineBreaks": 0,
    "offset": 103,
    "text": " ",
    "toString": [Function],
    "type": "WhiteSpace",
    "value": " ",
  },
  {
    "col": 105,
    "line": 1,
    "lineBreaks": 0,
    "offset": 104,
    "text": "options",
    "toString": [Function],
    "type": "AttributeName",
    "value": "options",
  },
  {
    "col": 112,
    "line": 1,
    "lineBreaks": 0,
    "offset": 111,
    "text": "=",
    "toString": [Function],
    "type": "Separator",
    "value": "=",
  },
  {
    "col": 113,
    "line": 1,
    "lineBreaks": 0,
    "offset": 112,
    "text": "'{\\"key\\": \\"value\\", \\"array\\": [1, 2, 3]}'",
    "toString": [Function],
    "type": "StringLiteral",
    "value": {
      "array": [
        1,
        2,
        3,
      ],
      "key": "value",
    },
  },
  {
    "col": 151,
    "line": 1,
    "lineBreaks": 0,
    "offset": 150,
    "text": " ",
    "toString": [Function],
    "type": "WhiteSpace",
    "value": " ",
  },
  {
    "col": 152,
    "line": 1,
    "lineBreaks": 0,
    "offset": 151,
    "text": "data-list",
    "toString": [Function],
    "type": "AttributeName",
    "value": "data-list",
  },
  {
    "col": 161,
    "line": 1,
    "lineBreaks": 0,
    "offset": 160,
    "text": "=",
    "toString": [Function],
    "type": "Separator",
    "value": "=",
  },
  {
    "col": 162,
    "line": 1,
    "lineBreaks": 0,
    "offset": 161,
    "text": "\\"[1, 2, 3]\\"",
    "toString": [Function],
    "type": "StringLiteral",
    "value": [
      1,
      2,
      3,
    ],
  },
  {
    "col": 173,
    "line": 1,
    "lineBreaks": 0,
    "offset": 172,
    "text": " ",
    "toString": [Function],
    "type": "WhiteSpace",
    "value": " ",
  },
  {
    "col": 174,
    "line": 1,
    "lineBreaks": 0,
    "offset": 173,
    "text": "punc",
    "toString": [Function],
    "type": "AttributeName",
    "value": "punc",
  },
  {
    "col": 178,
    "line": 1,
    "lineBreaks": 0,
    "offset": 177,
    "text": "=",
    "toString": [Function],
    "type": "Separator",
    "value": "=",
  },
  {
    "col": 179,
    "line": 1,
    "lineBreaks": 0,
    "offset": 178,
    "text": "\\"a=b,c,d,e\\"",
    "toString": [Function],
    "type": "StringLiteral",
    "value": "a=b,c,d,e",
  },
  {
    "col": 190,
    "line": 1,
    "lineBreaks": 0,
    "offset": 189,
    "text": " ",
    "toString": [Function],
    "type": "WhiteSpace",
    "value": " ",
  },
  {
    "col": 191,
    "line": 1,
    "lineBreaks": 0,
    "offset": 190,
    "text": "checked",
    "toString": [Function],
    "type": "AttributeName",
    "value": "checked",
  },
  {
    "col": 198,
    "line": 1,
    "lineBreaks": 0,
    "offset": 197,
    "text": "=",
    "toString": [Function],
    "type": "Separator",
    "value": "=",
  },
  {
    "col": 199,
    "line": 1,
    "lineBreaks": 0,
    "offset": 198,
    "text": "false",
    "toString": [Function],
    "type": "BooleanLiteral",
    "value": false,
  },
  {
    "col": 204,
    "line": 1,
    "lineBreaks": 0,
    "offset": 203,
    "text": " ",
    "toString": [Function],
    "type": "WhiteSpace",
    "value": " ",
  },
  {
    "col": 205,
    "line": 1,
    "lineBreaks": 0,
    "offset": 204,
    "text": "checked",
    "toString": [Function],
    "type": "AttributeName",
    "value": "checked",
  },
  {
    "col": 212,
    "line": 1,
    "lineBreaks": 0,
    "offset": 211,
    "text": "=",
    "toString": [Function],
    "type": "Separator",
    "value": "=",
  },
  {
    "col": 213,
    "line": 1,
    "lineBreaks": 0,
    "offset": 212,
    "text": "false",
    "toString": [Function],
    "type": "BooleanLiteral",
    "value": false,
  },
  {
    "col": 218,
    "line": 1,
    "lineBreaks": 0,
    "offset": 217,
    "text": " ",
    "toString": [Function],
    "type": "WhiteSpace",
    "value": " ",
  },
  {
    "col": 219,
    "line": 1,
    "lineBreaks": 0,
    "offset": 218,
    "text": "data-checked",
    "toString": [Function],
    "type": "AttributeName",
    "value": "data-checked",
  },
  {
    "col": 231,
    "line": 1,
    "lineBreaks": 0,
    "offset": 230,
    "text": "=",
    "toString": [Function],
    "type": "Separator",
    "value": "=",
  },
  {
    "col": 232,
    "line": 1,
    "lineBreaks": 0,
    "offset": 231,
    "text": "\\"false\\"",
    "toString": [Function],
    "type": "StringLiteral",
    "value": "false",
  },
  {
    "col": 239,
    "line": 1,
    "lineBreaks": 0,
    "offset": 238,
    "text": " ",
    "toString": [Function],
    "type": "WhiteSpace",
    "value": " ",
  },
  {
    "col": 240,
    "line": 1,
    "lineBreaks": 0,
    "offset": 239,
    "text": "disabled",
    "toString": [Function],
    "type": "AttributeName",
    "value": "disabled",
  },
]
`;
